### PR TITLE
EES-5172 Adding Admin endpoint to delete DRAFT API data set versions

### PR DESCRIPTION
**Pull request by jack-hive**
We've added a new endpoint to the Admin which will be used to delete DRAFT API data set versions, and all of their associated data. The deletion process consists of:
- Deleting the `DataSetVersion` entity from the Public Data DB
- Deleting the `DataSet` entity from the Public Data DB **IFF** the `DataSetVersion` being deleted, is the very first version
- Deleting ALL associated metadata with the `DataSetVersion`. This includes (each of which are found in the Public Data DB)*:
    - `FilterMetas` and their associated `FilterOptionMetaLinks`
    - `LocationMetas` and their associated `LocationOptionMetaLinks`
    - `GeographicLevelMetas`
    - `TimePeriodMetas`
    - `IndicatorMetas`
- Deleting the `DataSetVersionImport` (found in the Public Data DB) which will have been created for the `DataSetVersion` on initial creation*
- Unassociating the `File` in the Content DB with the public API data set
- Deleting ALL of the File Share data which was produced during the import process. This includes all Duck DB related files.

_*These are deleted automatically on the deletion of the `DataSetVersion` from the DB due to the cascade rules that are in place. Hence, I do not explicitly have to delete them in the code._

The endpoint only authenticates BAU users to trigger the deletion.

The endpoint will return a `400` if the `DataSetVersion` that is being attempted to be deleted is not in a `Draft` status.

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/DataSetVersionService.cs [hunk 1]
```diff
@@ -46,6 +65,29 @@ public async Task<List<DataSetVersionStatusSummary>> GetStatusesForReleaseVersio
...
+            return Unit.Instance;
+        }
+
+        return ValidationUtils.ValidationResult(new ErrorViewModel
+        {
+            Code = ValidationMessages.DataSetVersionCanNotBeDeleted.Code,
+            Message = ValidationMessages.DataSetVersionCanNotBeDeleted.Message,
+            Detail = new InvalidErrorDetail<Guid>(dataSetVersion.Id)
+        });
+    }
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/DataSetVersionService.cs:90]: I think we are duplicating this check in both the Admin and the Processor which isn't necessarily a problem but in https://github.com/dfe-analytical-services/explore-education-statistics/pull/4917 we have moved the Admin sides validation for creating a data set version to the Processor, so I'm thinking we should probably do the same here and make this nothing more than a proxy for the delete call. What do you think?

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Public.Data/DataSetVersionsControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+using Moq;
+using static GovUk.Education.ExploreEducationStatistics.Admin.Tests.Security.Utils.ClaimsPrincipalUtils;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Tests.Controllers.Api.Public.Data;
+
+public class DataSetVersionsControllerTests(TestApplicationFactory testApp) : IntegrationTestFixture(testApp)
+{
+    private const string BaseUrl = "api/public-data/data-sets/versions";
+
+    public class DeleteVersionTests(TestApplicationFactory testApp) : DataSetsControllerTests(testApp)
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Public.Data/DataSetVersionsControllerTests.cs:28]: I think this should be `: DataSetVersionsControllerTests(testApp)`, not `: DataSetsControllerTests(testApp)`.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Public.Data/DataSetVersionsControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+using GovUk.Education.ExploreEducationStatistics.Public.Data.Model.Tests.Fixtures;
+using GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Requests.Validators;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Testing;
+using Moq;
+using static GovUk.Education.ExploreEducationStatistics.Admin.Tests.Security.Utils.ClaimsPrincipalUtils;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Tests.Controllers.Api.Public.Data;
+
+public class DataSetVersionsControllerTests(TestApplicationFactory testApp) : IntegrationTestFixture(testApp)
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Public.Data/DataSetVersionsControllerTests.cs:24]: I think this class could be `abstract`?

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Public.Data/DataSetVersionsControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+                                }
+                        })));
+
+            var client = BuildApp(processorClient.Object).CreateClient();
+
+            var response = await DeleteVersion(dataSetVersion.Id, client);
+
+            var validationProblem = response.AssertValidationProblem();
+
+            var error = validationProblem.AssertHasError(null, Errors.Error1.ToString());
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Public.Data/DataSetVersionsControllerTests.cs:232]: As these are both string params I think it would be a good idea to label them, and it would give an indication to the reader to say what this `null` represents (i.e. the path) without having to inspect it.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests/Functions/DeleteDataSetVersionFunctionTests.cs [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+
+            releaseFile.File.PublicApiDataSetId = dataSet.Id;
+            releaseFile.File.PublicApiDataSetVersion = dataSetVersion.FullSemanticVersion();
+
+            await AddTestData<ContentDbContext>(context => context.Files.Update(releaseFile.File));
+
+            var dataSetVersionDirectory = _dataSetVersionPathResolver.DirectoryPath(dataSetVersion);
+
+            Directory.CreateDirectory(dataSetVersionDirectory);
+            System.IO.File.WriteAllText(Path.Combine(dataSetVersionDirectory, "version1.txt"), "dummy file text");
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests/Functions/DeleteDataSetVersionFunctionTests.cs:80]: I think you can use `WriteAllTextAsync` here. Alterntively you could create an empty file with

```csharp
await System.IO.File.Create(Path.Combine(dataSetVersionDirectory, "version1.txt")).DisposeAsync();
```

as we aren't particularly bothered about the content.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests/Functions/DeleteDataSetVersionFunctionTests.cs [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+                    .Where(fm => fm.DataSetVersionId == dataSetVersion.Id)
+                    .CountAsync());
+            Assert.Equal(0,
+                await publicDataDbContext.GeographicLevelMetas
+                    .Where(fm => fm.DataSetVersionId == dataSetVersion.Id)
+                    .CountAsync());
+            Assert.Equal(0,
+                await publicDataDbContext.TimePeriodMetas
+                    .Where(fm => fm.DataSetVersionId == dataSetVersion.Id)
+                    .CountAsync());
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests/Functions/DeleteDataSetVersionFunctionTests.cs:125]: Might be a little more concise to use `AnyAsync` here instead, like

```csharp
            Assert.False(await publicDataDbContext.FilterMetas
                    .AnyAsync(fm => fm.DataSetVersionId == dataSetVersion.Id));
```

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/Public.Data/DataSetVersionsController.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+using GovUk.Education.ExploreEducationStatistics.Admin.Services.Interfaces.Public.Data;
+using GovUk.Education.ExploreEducationStatistics.Common.Extensions;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Controllers.Api.Public.Data;
+
+[Authorize]
+[ApiController]
+[Route("api/public-data/data-sets/versions")]
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/Public.Data/DataSetVersionsController.cs:14]: Would suggest going with `api/public-data/data-set-versions` for the URL

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/ProcessorClient.cs [hunk 8]
```diff
@@ -59,4 +59,36 @@ await response.Content
...
+                    return new NotFoundResult();
+                default:
+                    var message = await response.Content.ReadAsStringAsync(cancellationToken);
+                    logger.LogError($"""
+                         Failed to delete data set version with status code: {response.StatusCode}. Message:
+                         {message}
+                         """);
+                    response.EnsureSuccessStatusCode();
+                    break;
+            }
```
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/ProcessorClient.cs:89]: I did think of doing this, however I decided against it because they're distinctly different endpoint calls and have no reason to necessarily be handled the same way or have the same status codes to handle. It's just coincidence in this case that they're the same. In fact, originally, the processor didn't return 400s so this method didn't handle this case so was different to the `CreateInitialDataSetVersion` method. It then became identical when I added 400s to the processor endpoint. But like I say, this could easily change and become different again.

So, I'm wondering whether to leave them separate on purpose, even if there is some repetition?

One thing to consider is, I will be adding another method to this interface when I add the bulk delete endpoint to the processor, and it may respond with the same status codes again, and therefore look identical to these again. Which may sway me to do what you've said since at that point we'd have 3 methods that could reuse a lot of stuff.

In summary, I'm undecided because they feel like they shouldn't be made to try to share stuff, but the fact is they're pretty identical.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests.csproj [hunk 9]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests.csproj:1]: I'm confused about what I actually changed? 😅

[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests.csproj]: fixed :)

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Functions/DeleteDataSetVersionFunction.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+using Microsoft.Azure.Functions.Worker;
+using Microsoft.Extensions.Logging;
+
+namespace GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Functions;
+
+public class DeleteDataSetVersionFunction(IDataSetVersionService dataSetVersionService, ILogger<DeleteDataSetVersionFunction> logger)
+{
+    [Function(nameof(DeleteDataSetVersion))]
+    public async Task<IActionResult> DeleteDataSetVersion(
+        [HttpTrigger(AuthorizationLevel.Function, "delete", Route = $"{nameof(DeleteDataSetVersion)}/{{dataSetVersionId}}")] HttpRequest httpRequest,
```
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Functions/DeleteDataSetVersionFunction.cs:14]: done :)

[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Functions/DeleteDataSetVersionFunction.cs:14]: It doesn't appear like it unfortunately. I get a binding error when I don't specify a `HttpRequest httpRequest` parameter. It also needs to be named `httpRequest` exactly. Looks like a weird quirk built into the framework.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Functions/DeleteDataSetVersionFunction.cs [hunk 11]
```diff
@@ -0,0 +1,31 @@
+using GovUk.Education.ExploreEducationStatistics.Common.Extensions;
+using GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Services.Interfaces;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Azure.Functions.Worker;
+using Microsoft.Extensions.Logging;
+
+namespace GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Functions;
+
+public class DeleteDataSetVersionFunction(IDataSetVersionService dataSetVersionService, ILogger<DeleteDataSetVersionFunction> logger)
```
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Functions/DeleteDataSetVersionFunction.cs:10]: reformatted :)

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Functions/DeleteDataSetVersionFunction.cs [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        try
+        {
+            return await dataSetVersionService.DeleteVersion(
+                    dataSetVersionId,
+                    cancellationToken: cancellationToken)
+                .HandleFailuresOrNoContent(convertNotFoundToNoContent: false);
+        }
+        catch (Exception ex)
+        {
+            logger.LogError(exception: ex, null, []);
```
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Functions/DeleteDataSetVersionFunction.cs:27]: Yeah that's probably better!

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionService.cs [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        if (dataSetVersion.IsFirstVersion)
+        {
+            var dataSetDirectory = Directory.GetParent(directory)!.FullName;
+
+            Directory.Delete(dataSetDirectory, true);
+
+            return;
+        }
+
+        Directory.Delete(directory, true);
```
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionService.cs:118]: I wonder though, in this case, we'd be deleting the child directory and then deleting the parent directory afterwards (both recursively)

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionService.cs [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    {
+        publicDataDbContext.DataSetVersions.Remove(dataSetVersion);
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+
+        if (dataSetVersion.IsFirstVersion)
+        {
+            publicDataDbContext.DataSets.Remove(dataSetVersion.DataSet);
+            await publicDataDbContext.SaveChangesAsync(cancellationToken);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionService.cs:84]: I think the only way is to save the deletion of the `DataSetVersion` first, but I may be wrong

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionService.cs [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+namespace GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Services;
+
+internal class DataSetVersionService(
+    ContentDbContext contentDbContext,
+    PublicDataDbContext publicDataDbContext,
+    IDataSetVersionPathResolver dataSetVersionPathResolver
+    ) : IDataSetVersionService
+{
+    public async Task<Either<ActionResult, Unit>> DeleteVersion(Guid dataSetVersionId, CancellationToken cancellationToken = default)
```
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionService.cs:25]: I've reformatted this file too :)

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Public.Data/DataSetVersionsControllerTests.cs (hunk 2 above)
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Public.Data/DataSetVersionsControllerTests.cs:28]: good spot!

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Public.Data/DataSetVersionsControllerTests.cs (hunk 4 above)
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Public.Data/DataSetVersionsControllerTests.cs:232]: I've just gotten rid of the `Enum` and hardcoded some strings for the `Code` and `Path` :)

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests/Functions/DeleteDataSetVersionFunctionTests.cs (hunk 5 above)
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests/Functions/DeleteDataSetVersionFunctionTests.cs:80]: good shout, I've gone for just creating an empty file

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests/Functions/DeleteDataSetVersionFunctionTests.cs (hunk 6 above)
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Tests/Functions/DeleteDataSetVersionFunctionTests.cs:125]: much better!

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/DataSetVersionService.cs (hunk 1 above)
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/DataSetVersionService.cs:90]: I've stopped the admin from doing any other checks than checking they are a BAU user. I've adjusted the integration tests to reflect this too

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionService.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionService.cs:89]: Ok no worries. I'm still getting to grips with how the relationships are configured and enforced so I was hoping one of those might have worked! 🙂

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionService.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionService.cs:124]: Ok yeah, not sure what I was thinking here now! I think I'd misread the code.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/ProcessorClient.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/ProcessorClient.cs:101]: Ok yep I thought this might be the case. I agree with your point about handling them both in a distinct way for now, and maybe this decision changes when the next operation gets added to it.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Functions/DeleteDataSetVersionFunction.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Functions/DeleteDataSetVersionFunction.cs]: Ok cool, that's disappointing.

One post I saw suggested using `string dataSetVersionId` and then parsing it to Guid later might work, but it doesn't seem to unfortunately.

While I was trying it out myself I found that it doesn't have to be named `httpRequest` as in my test case I was using `HttpRequest request` and this was ok. This makes sense because our `HealthCheckFunctions.HealthCheck` is already using `HttpRequest request` too.

One thing you might want to change though is constraining the type of the `dataSetVersionId` parameter to Guid using

```csharp
[HttpTrigger(AuthorizationLevel.Anonymous, "delete", Route=${nameof(DeleteDataSetVersion)}/{{dataSetVersionId:guid}}")]
```

If you do this and the value is not a valid Guid then it doesn't match the route.

**Review by benoutram — APPROVED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Functions/DeleteDataSetVersionFunction.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Functions/DeleteDataSetVersionFunction.cs]: Ah yeah, now that you mention it I vaguely remember reading that there are a few allowed names for that variable. `request` may have been one of them.

I'll change it to `request` and I'll add the `:guid` constraint, nice spot!

**Review by benoutram — APPROVED** [review]
